### PR TITLE
Region tags for standalone snippets.

### PR DESCRIPTION
**Pull request by amanda-tarafa**
As always, a commit at a time.

(The two last commits may be squashed before merging)

### CODE REVIEW

**Comment by amanda-tarafa** [comment]
(Checking the test failure)

**Comment by amanda-tarafa** [comment]
(Test error fixed)

File: Google.Api.Generator.Tests/ProtoTest.cs [hunk 1]
```diff
@@ -92,18 +92,24 @@ public void ProtocExecution()
             }
         }
 
-        private void BuildTest(string testName, bool ignoreUnitTests = false)
+        private void BuildTest(string serviceName, string protoPackageVersion = null, bool ignoreUnitTests = false)
         {
-            var baseTestPath = Path.Combine(Invoker.GeneratorTestsDir, "ProtoTests", testName);
-            var clientPath = Path.Combine(baseTestPath, $"Testing.{testName}");
+            var (effectiveBasePathPart, effectiveTestName) = protoPackageVersion == null 
+                ? ( serviceName, serviceName)
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTest.cs:98]: Nit: remove space after `(`

File: Google.Api.Generator.Tests/ProtoTest.cs [hunk 2]
```diff
@@ -92,18 +92,24 @@ public void ProtocExecution()
...
         }
 
-        private void BuildTest(string testName, bool ignoreUnitTests = false)
+        private void BuildTest(string serviceName, string protoPackageVersion = null, bool ignoreUnitTests = false)
         {
-            var baseTestPath = Path.Combine(Invoker.GeneratorTestsDir, "ProtoTests", testName);
-            var clientPath = Path.Combine(baseTestPath, $"Testing.{testName}");
+            var (effectiveBasePathPart, effectiveTestName) = protoPackageVersion == null 
+                ? ( serviceName, serviceName) 
+                : ($"{serviceName}.{protoPackageVersion}", $"{serviceName}.{protoPackageVersion.Replace('v', 'V')}");
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTest.cs:99]: This won't work if we have v1beta - we'd want V1Beta, not V1beta. Maybe just upper-case `protoPackageVersion`?

File: Google.Api.Generator.Tests/ProtoTests/Basic.v1/Testing.Basic.V1.StandaloneSnippets/Testing.Basic.V1.StandaloneSnippets.csproj [hunk 3]
```diff
@@ -0,0 +1,15 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/Basic.v1/Testing.Basic.V1.StandaloneSnippets/Testing.Basic.V1.StandaloneSnippets.csproj:1]: I'd expect the project files to be renames too - perhaps you didn't delete them from Testing.Basic (or didn't add the delete to Git?)

File: Google.Api.Generator.Utils/Formatting/WhitespaceFormatter.cs [hunk 4]
```diff
@@ -133,15 +140,26 @@ public override SyntaxNode VisitNamespaceDeclaration(NamespaceDeclarationSyntax
             using (WithIndent())
             {
                 node = (NamespaceDeclarationSyntax)base.VisitNamespaceDeclaration(node);
+
+                // Probably the start region tag in snippets.
```
Comment (jskeet) [inline Google.Api.Generator.Utils/Formatting/WhitespaceFormatter.cs:144]: I don't understand what this means. Are you saying we don't get here at all unless there are region tags?

File: Google.Api.Generator.Utils/Formatting/WhitespaceFormatter.cs [hunk 5]
```diff
@@ -78,6 +78,13 @@ private IDisposable WithIndent(bool withIndent = true)
         private SyntaxTriviaList FormatXmlDoc(SyntaxTriviaList trivList) =>
             XmlDocSplitter.Split(_indentTrivia, _maxLineLength, trivList);
 
+        private IEnumerable<SyntaxTrivia> AdjustTrivia(SyntaxTrivia triv) => triv switch
```
Comment (jskeet) [inline Google.Api.Generator.Utils/Formatting/WhitespaceFormatter.cs:81]: I think this is work a comment. I can't immediately guess what this is trying to do.

File: Google.Api.Generator/Generation/SnippetCodeGenerator.cs [hunk 6]
```diff
@@ -169,7 +172,30 @@ public NamespaceDeclarationSyntax StandaloneSnippet(SourceFileContext ctx)
...
+            {
+                string effectiveShortName = TargetMethod.Svc.DefaultHostServiceName ?? "unknown";
+                string effectiveVersion = TargetMethod.Svc.PackageVersion is null ? "" : $"_{TargetMethod.Svc.PackageVersion}";
+                string syncText = Sync ? "sync" : "async";
+                string effectiveDisambiguation = RegionTagDisambiguation is null ? "" : $"_{RegionTagDisambiguation}";
+
+                // {apishortname}_{apiVersion}_generated_{serviceName}_{rpcName}_{sync|async}_{disambiguation}
+                string regionTagName =
+                    $"{effectiveShortName}" +
+                    $"{effectiveVersion}" +
```
Comment (jskeet) [inline Google.Api.Generator/Generation/SnippetCodeGenerator.cs:189]: Aha... I'd been wondering why we actually need the version...

File: Google.Api.Generator/Generation/SnippetCodeGenerator.cs [hunk 7]
```diff
@@ -169,7 +172,30 @@ public NamespaceDeclarationSyntax StandaloneSnippet(SourceFileContext ctx)
...
+
+            private NamespaceDeclarationSyntax WithRegionTag(NamespaceDeclarationSyntax ns)
+            {
+                string effectiveShortName = TargetMethod.Svc.DefaultHostServiceName ?? "unknown";
+                string effectiveVersion = TargetMethod.Svc.PackageVersion is null ? "" : $"_{TargetMethod.Svc.PackageVersion}";
+                string syncText = Sync ? "sync" : "async";
+                string effectiveDisambiguation = RegionTagDisambiguation is null ? "" : $"_{RegionTagDisambiguation}";
+
+                // {apishortname}_{apiVersion}_generated_{serviceName}_{rpcName}_{sync|async}_{disambiguation}
+                string regionTagName =
```
Comment (jskeet) [inline Google.Api.Generator/Generation/SnippetCodeGenerator.cs:187]: Maybe unwrap all of this onto one line? I think that would make it easier to correlate it with the comment.

File: Google.Api.Generator.Tests/ProtoTests/Paginated/Testing.Paginated.StandaloneSnippets/PaginatedClient.ResourcedMethod2Snippet.g.cs [hunk 8]
```diff
@@ -16,6 +16,7 @@
 
 namespace Testing.Paginated.Snippets
 {
+    // [START paginated_generated_Paginated_ResourcedMethod_sync_flattened2]
```
Comment (jskeet) [inline Google.Api.Generator.Tests/ProtoTests/Paginated/Testing.Paginated.StandaloneSnippets/PaginatedClient.ResourcedMethod2Snippet.g.cs:19]: Where we've got disambiguation like this, does it matter if the index isn't stable over time?

**Review by jskeet — APPROVED** [review]
Various comments, but basically fine.

File: Google.Api.Generator.Tests/ProtoTests/Basic.v1/Testing.Basic.V1.StandaloneSnippets/Testing.Basic.V1.StandaloneSnippets.csproj (hunk 3 above)
Comment (amanda-tarafa) [reply Google.Api.Generator.Tests/ProtoTests/Basic.v1/Testing.Basic.V1.StandaloneSnippets/Testing.Basic.V1.StandaloneSnippets.csproj]: Didn't add the delete to Git. Done now.

File: Google.Api.Generator.Utils/Formatting/WhitespaceFormatter.cs (hunk 5 above)
Comment (amanda-tarafa) [reply Google.Api.Generator.Utils/Formatting/WhitespaceFormatter.cs]: Changed the name of the method and added comments.

File: Google.Api.Generator.Utils/Formatting/WhitespaceFormatter.cs (hunk 4 above)
Comment (amanda-tarafa) [reply Google.Api.Generator.Utils/Formatting/WhitespaceFormatter.cs]: We get here every time.
I meant that if the trivia is not empty, is probably the region tag.
Before adding the region tag this was just being replaced by indents and new lines, so it was assumed that it was empty.
I've reworded the comment.

File: Google.Api.Generator/Generation/SnippetCodeGenerator.cs (hunk 7 above)
Comment (amanda-tarafa) [reply Google.Api.Generator/Generation/SnippetCodeGenerator.cs:186]: Done

File: Google.Api.Generator/Generation/SnippetCodeGenerator.cs (hunk 6 above)
Comment (amanda-tarafa) [reply Google.Api.Generator/Generation/SnippetCodeGenerator.cs]: Yep :)

File: Google.Api.Generator.Tests/ProtoTests/Paginated/Testing.Paginated.StandaloneSnippets/PaginatedClient.ResourcedMethod2Snippet.g.cs (hunk 8 above)
Comment (amanda-tarafa) [reply Google.Api.Generator.Tests/ProtoTests/Paginated/Testing.Paginated.StandaloneSnippets/PaginatedClient.ResourcedMethod2Snippet.g.cs:19]: Unfortunatelly yes, it should be stable.
If we get all systems to depend on metadata only, then we won't care about the region tag that much, but that's not the case right now (although I'm trying).
That said, in reality (and this applies for all languages), these samples will most likely be included in reference docs only, either via CloudRad or API reference docs, and none of that will depend on the region tag.
I don't think these samples will be broadly consumed by TWs, which is what the region tag is ultimately useful for.

**Review by amanda-tarafa — COMMENTED** [review]
All comments addressed!

File: Google.Api.Generator.Utils/Formatting/WhitespaceFormatter.cs [hunk 9]
```diff
@@ -133,15 +148,26 @@ public override SyntaxNode VisitNamespaceDeclaration(NamespaceDeclarationSyntax
             using (WithIndent())
             {
                 node = (NamespaceDeclarationSyntax)base.VisitNamespaceDeclaration(node);
+
+                // If the trivia is not empty, it's probablt the start region tag.
```
Comment (jskeet) [inline Google.Api.Generator.Utils/Formatting/WhitespaceFormatter.cs:152]: Much clearer, but now with a tiny typo - probablt => probably :) (Same on 162)

**Review by jskeet — APPROVED** [review]

**Pull request merged**